File: 70.py
#!/usr/bin/env python

"""This script solves the Project Euler problem "Totient permutation". The
problem is: Find the value of n, 1 < n < 10^7, for which phi(n) is a
permutation of n and the ratio n/phi(n) produces a minimum.
"""

from __future__ import division
import math


def main():
    """Totient permutation"""

    # Constants
    LIMIT = 1e7

    min_ratio = None
    n_for_min_ratio = None

    primes = get_primes_up_to(int(2 * math.sqrt(LIMIT)))

    for i in primes:
        for j in primes:
            if j < i:
                continue
            n = i * j
            if n >= LIMIT:
                continue
            phi = n * (1 - 1 / i)
            if i != j:
                phi *= (1 - 1 / j)
            if not is_perm(n, int(phi)):
                continue
            ratio = n / phi
            if min_ratio is None or ratio < min_ratio:
                min_ratio = ratio
                n_for_min_ratio = n

    print(n_for_min_ratio)


def get_primes_up_to(limit):
    """Get all primes up to specified limit"""

    sieve_bound = (limit - 1) // 2  # Last index of sieve
    sieve = [False for _ in range(sieve_bound)]
    cross_limit = (math.sqrt(limit) - 1) // 2

    i = 1
    while i <= cross_limit:
        if not sieve[i - 1]:
            # 2 * $i + 1 is prime, so mark multiples
            j = 2 * i * (i + 1)
            while j <= sieve_bound:
                sieve[j - 1] = True
                j += 2 * i + 1
        i += 1

    primes = [2 * n + 1 for n in range(1, sieve_bound + 1) if not sieve[n - 1]]
    primes.insert(0, 2)

    return(primes)


def is_perm(num1, num2):
    """Check if two numbers are permutations of each other"""

    return(sorted(str(num1)) == sorted(str(num2)))

if __name__ == '__main__':
    main()

File: 70.R
# Totient permutation

# Constants
LIMIT <- 1e7

get_primes_up_to <- function(limit) {
    sieve_bound <- (limit - 1) %/% 2 # Last index of sieve
    sieve <- rep(FALSE, sieve_bound)
    cross_limit <- (floor(sqrt(limit)) - 1) %/% 2

    i <- 1
    while ( i <= cross_limit ) {
        if (!sieve[i]) {
            # 2 * $i + 1 is prime, so mark multiples
            j <- 2 * i * (i + 1)
            while ( j <= sieve_bound ) {
                sieve[j] <- TRUE
                j <- j + 2 * i + 1
            }
        }
        i <- i + 1
    }

    primes <- rep(0, sieve_bound)
    for ( i in seq(1, sieve_bound) ) {
        if ( !sieve[i] ) {
            primes[i] <- 2 * i + 1
        }
    }

    return(c(2, primes[primes > 0]))
}

sort_digits <- function(num) {
    return(paste(sort(strsplit(as.character(num), split=NULL)[[1]]),
                 collapse=""))
}

primes <- get_primes_up_to(2 * sqrt(LIMIT))

df <- expand.grid(prime1=primes, prime2=primes)
df$n <- df$prime1 * df$prime2
df$phi <- df$n * (1 - 1 / df$prime1) * (1 - 1 / df$prime2)
df$ratio <- df$n / df$phi

df$nsorted <- sapply(df$n, sort_digits)
df$phisorted <- sapply(df$phi, sort_digits)

min_ratio <- min(df$ratio[df$nsorted == df$phisorted & df$n < LIMIT])
n_for_min_ratio <- df$n[df$ratio == min_ratio]

cat(n_for_min_ratio[1], fill=TRUE)
